sdk/identity/azure_identity/src/lib.rs

```diff
@@ -134,30 +161,48 @@ fn get_authority_host(env: Option<Env>, cloud: Option<&CloudConfiguration>) -> R
 }
 
 const TSG_LINK_ERROR_TEXT: &str =
-    ". To troubleshoot, visit https://aka.ms/azsdk/rust/identity/troubleshoot";
+    "\nTo troubleshoot, visit https://aka.ms/azsdk/rust/identity/troubleshoot";
 
 /// Map an error from a credential's get_token() method to an ErrorKind::Credential error, appending
 /// a link to the troubleshooting guide entry for that credential, if it has one.
-///
-/// TODO: decide whether to map to ErrorKind::Credential here (https://github.com/Azure/azure-sdk-for-rust/issues/3127)
-fn authentication_error<T: 'static>(e: azure_core::Error) -> azure_core::Error {
-    azure_core::Error::with_message_fn(e.kind().clone(), || {
-        let type_name = std::any::type_name::<T>();
-        let short_name = type_name.rsplit("::").next().unwrap_or(type_name); // cspell:ignore rsplit
-        let link = match short_name {
-            "AzureCliCredential" => format!("{TSG_LINK_ERROR_TEXT}#azure-cli"),
-            "AzureDeveloperCliCredential" => format!("{TSG_LINK_ERROR_TEXT}#azd"),
-            "AzurePipelinesCredential" => format!("{TSG_LINK_ERROR_TEXT}#apc"),
-            #[cfg(feature = "client_certificate")]
-            "ClientCertificateCredential" => format!("{TSG_LINK_ERROR_TEXT}#client-cert"),
-            "ClientSecretCredential" => format!("{TSG_LINK_ERROR_TEXT}#client-secret"),
-            "ManagedIdentityCredential" => format!("{TSG_LINK_ERROR_TEXT}#managed-id"),
-            "WorkloadIdentityCredential" => format!("{TSG_LINK_ERROR_TEXT}#workload"),
-            _ => "".to_string(),
-        };
-
-        format!("{short_name} authentication failed: {e}{link}")
-    })
+fn authentication_error<T>(err: azure_core::Error) -> azure_core::Error {
+    let type_name = std::any::type_name::<T>();
+    // remove generic parameters; we want only the credential name
+    let type_name = type_name.split('<').next().unwrap_or(type_name);
+    let short_name = type_name.rsplit("::").next().unwrap_or(type_name); // cspell:ignore rsplit
+
+    let (wrap_inner_error, link_fragment) = match short_name {
+        stringify!(AzureCliCredential) => (false, "#azure-cli"),
+        stringify!(AzureDeveloperCliCredential) => (false, "#azd"),
+        stringify!(AzurePipelinesCredential) => (true, "#apc"),
+        stringify!(ClientCertificateCredential) => (false, "#client-cert"),
+        stringify!(ClientSecretCredential) => (false, "#client-secret"),
+        stringify!(ManagedIdentityCredential) => (false, "#managed-id"),
+        stringify!(WorkloadIdentityCredential) => (true, "#workload"),
+        _ => (false, ""),
+    };
+
+    let mut details = err.to_string();
+    if wrap_inner_error {
+        // details is e.g. "ClientAssertionCredential authentication failed. <more details>".
+        // We want only the "<more details>" part because we want to return an error like
+        // "OtherCredential authentication failed. <more details>".
```

Comment:
Not blocking, but I'd urge a cheaper way in the future. If you could do something in `err` to tell it came from an internal implementation detail or even pass something down to `ClientAssertionCredential` to tell it which name to use, you could avoid all this fragile and (slightly) time-wasting string manipulation.